rfd/0011-database-access.md

```diff
@@ -0,0 +1,698 @@
+postgres     PostgreSQL 13.0 env=dev
+postgres-rds PostgreSQL 12.4 env=stage
+```
+
+To log into a specific database instance, a user executes `tsh db login`
+command. This will retrieve the certificate with encoded database information
+in it:
+
+```sh
+$ tsh db login postgres
+```
+
+The `tsh status` command shows the name of the database a user is logged into:
+
+```sh
+$ tsh status
+> Profile URL:        https://127.0.0.1:3080
+  Logged in as:       alice
+  Roles:              db-developer*
+  ...
+  Databases:          postgres
+  ...
+```
+
+User can be logged into multiple databases simultaneously:
+
+```sh
+$ tsh db login db1
+$ tsh db login db3
+$ tsh db ls
+Name  Description Labels
+----- ----------- ---------
+> db1 Database 1  env=dev
+db2   Database 2  env=stage
+> db3 Database 3  env=prod
+```
+
+To log out of a particular database (i.e. remove certificate from key store):
+
+```sh
+$ tsh db logout db1
+```
+
+### Selecting databases and database user
+
+The `tsh db login` command also prints a footer explaining how to connect
+to the database:
+
+```
+Connection information for "postgres" has been saved to ~/.pg_service.conf.
+You can connect to the database using the following command:
+
+  $ psql "service=postgres user=<user> dbname=<dbname>"
+
+Or configure environment variables and use regular CLI flags:
+
+  $ eval $(tsh db env)
+  $ psql -U <user> <database>
+```
+
+After a successful login, users can use any database user and database to
+connect to, as long as it is allowed by the RBAC rules. For example:
+
+```sh
+$ tsh db login postgres-prod && eval $(tsh db env)
+$ psql -U alice mydb     # allowed by role so will connect
+$ psql -U bob mydb       # allowed by role so will connect
+$ psql -U charlie mydb   # not allowed by role so will deny access
+```
+
+Default database user and name can also be provided to the `tsh db login`
+command, in which case then won't be required when using `psql` or other
+client:
+
+```sh
+$ tsh db login --db-user=alice --db-name=mydb postgres-prod
+$ psql "service=postgres-prod"
+$ eval $(tsh db env)
+$ psql
+```
+
+Users can still choose to specify a user/database explicitly when connecting,
+for example to use a different database user or connect to a different 
+database (subject to RBAC checks).
+
+### Connecting to PostgreSQL / service file
+
+After logging into a PostgreSQL database, `tsh` configures an appropriate
+section in the [connection service file](https://www.postgresql.org/docs/9.1/libpq-pgservice.html)
+which is located at `~/.pg_service.conf`.
+
+The service file has ini format and can contain multiple sections where each
+section defines connection parameters for a particular "service". PostgreSQL
+clients can refer to a particular service using `service` connection string
+parameter which all libpq-based clients should recognize (e.g. `psql` and
+`pgAdmin` do).
+
+The section is named after the Teleport database service instance a user has
```

Comment:
for k8s, we prefix the name with teleport cluster name (`${TELEPORT_CLUSTER}-${KUBE_CLUSTER}`) to deal with trusted clusters using overlapping db names
it might be useful here too

Reply:
Yeah, good point - I've been thinking about the same, will update here and in the implementation. Will worsen the UX a little bit but I don't see other good solution for potential name conflicts...